lib/DxilValidation/DxilContainerValidation.cpp

```diff
@@ -73,6 +78,610 @@ struct DiagRestore {
+  // Find the signature element in the set.
+  PSVSignatureElement0 PSVSE0;
+  InitPSVSignatureElement(PSVSE0, SE, i1ToUnknownCompat);
+
+  PSVSE0.SemanticIndexes =
+      findSemanticIndex(IndexTab, SE.GetSemanticIndexVec());
+
+  auto it = SigSet.find(&PSVSE0);
+  if (it == SigSet.end()) {
+    PSVSignatureElement PSVSE(StrTab, IndexTab, &PSVSE0);
+    std::string Str;
+    raw_string_ostream OS(Str);
+    PSVSE.Print(OS, SE.GetName());
+    OS.flush();
+    EmitMissingError(Name + "Element", Str);
+    return;
+  }
+  // Check the Name and SemanticIndex.
+  bool Mismatch = false;
+  PSVSignatureElement0 *PSVSE0Found = *it;
+  PSVSignatureElement PSVSEFound(StrTab, IndexTab, PSVSE0Found);
+  if (SE.IsArbitrary())
+    Mismatch = strcmp(PSVSEFound.GetSemanticName(), SE.GetName());
+  else
+    Mismatch = PSVSE0Found->SemanticKind != static_cast<uint8_t>(SE.GetKind());
+
+  PSVSE0.SemanticName = PSVSE0Found->SemanticName;
+  // Compare all fields.
+  Mismatch |= memcmp(&PSVSE0, PSVSE0Found, sizeof(PSVSignatureElement0)) != 0;
+  if (Mismatch) {
+    PSVSignatureElement PSVSE(StrTab, IndexTab, &PSVSE0);
+    std::string Str = GetDump(PSVSE);
+    std::string Str1;
+    raw_string_ostream OS(Str1);
+    PSVSEFound.Print(OS, SE.GetName());
+    OS.flush();
+    EmitMismatchError(Name + "Element", Str, Str1);
+  }
+}
+
+template <typename T>
+void PSVContentVerifier::VerifyResourceTable(T &ResTab, PSVResourceSet &ResSet,
+                                             unsigned PSVVersion) {
+  for (auto &&R : ResTab) {
+    PSVResourceBindInfo1 BI;
+    InitPSVResourceBinding(&BI, &BI, R.get());
+    auto It = ResSet.find(&BI);
+    if (It == ResSet.end()) {
+      std::string Str = GetDump(BI);
+      EmitMissingError("ResourceBindInfo", Str);
+      return;
+    }
+
+    if (PSVVersion > 1) {
+      PSVResourceBindInfo1 *BindInfo1 = (PSVResourceBindInfo1 *)*It;
+      if (memcmp(&BI, BindInfo1, sizeof(PSVResourceBindInfo1)) != 0) {
+        std::string Str = GetDump(BI);
+        std::string Str1 = GetDump(*BindInfo1);
+        EmitMismatchError("ResourceBindInfo", Str, Str1);
+      }
+    } else {
+      PSVResourceBindInfo0 *BindInfo = *It;
+      if (memcmp(&BI, BindInfo, sizeof(PSVResourceBindInfo0)) != 0) {
+        std::string Str = GetDump(BI);
+        std::string Str1 = GetDump(*BindInfo);
+        EmitMismatchError("ResourceBindInfo", Str, Str1);
+      }
+    }
+  }
+}
+
+void PSVContentVerifier::VerifyResources(unsigned PSVVersion) {
+  UINT uCBuffers = DM.GetCBuffers().size();
+  UINT uSamplers = DM.GetSamplers().size();
+  UINT uSRVs = DM.GetSRVs().size();
+  UINT uUAVs = DM.GetUAVs().size();
+  unsigned ResourceCount = uCBuffers + uSamplers + uSRVs + uUAVs;
+  if (PSV.GetBindCount() != ResourceCount) {
+    EmitMismatchError("ResourceCount", std::to_string(PSV.GetBindCount()),
+                      std::to_string(ResourceCount));
+    return;
+  }
+  // Build PSVResourceSet with data in PSV.
+  PSVResourceSet ResBindInfo0Set;
```

Comment:
What's the justification for relaxing the ordering requirement of resources defined in PSV0 in relation to DxilModule resources?  I don't think this should be allowed.

Reply:
Fixed.